collector/logs/sources/tail/tailer.go

```diff
@@ -0,0 +1,201 @@
+
+	return tailer, nil
+}
+
+// Stop stops the tailer and cleans up resources.
+// Does not wait for the tailer to finish processing, to allow closing many tailers concurrently.
+// Call Wait() after calling Stop() to wait for the tailer to finish processing.
+func (t *Tailer) Stop() {
+	t.tail.Cleanup()
+	t.tail.Stop()
+	t.shutdown()
+}
+
+// Wait waits for the tailer to finish processing.
+func (t *Tailer) Wait() {
+	t.errgroup.Wait()
+}
+
+func readLines(ctx context.Context, tailer *Tailer, updateChannel <-chan FileTailTarget, outputQueue chan<- *types.Log) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		// Receive updates from the optional updateChannel.
+		case newTarget, ok := <-updateChannel:
+			if ok {
+				newParsers, err := parser.NewParsers(newTarget.Parsers)
+				if err != nil {
+					logger.Errorf("readLines: parser error for filename %q: %v", tailer.tail.Filename, err)
+					continue
+				}
+				tailer.logLineParsers = newParsers
+				tailer.database = newTarget.Database
+				tailer.table = newTarget.Table
+				tailer.resources = make(map[string]interface{})
+				for k, v := range newTarget.Resources {
+					tailer.resources[k] = v
+				}
+			}
+		case line, ok := <-tailer.tail.Lines:
+			if !ok {
+				logger.Infof("readLines: tailer closed the channel for filename %q", tailer.tail.Filename)
+				return nil // No longer getting lines due to the tailer being closed. Exit.
+			}
+			if line.Err != nil {
+				logger.Errorf("readLines: tailer error for filename %q: %v", tailer.tail.Filename, line.Err)
+				//skip
+				continue
+			}
+
+			log := types.LogPool.Get(1).(*types.Log)
+			log.Reset()
+
+			isPartial, err := tailer.logTypeParser.Parse(line.Text, log)
+			if err != nil {
+				logger.Errorf("readLines: parselog error for filename %q: %v", tailer.tail.Filename, err)
+				//skip
+				types.LogPool.Put(log)
+				continue
+			}
+			if isPartial {
+				types.LogPool.Put(log)
+				continue
+			}
+
+			position := line.Offset
+			currentFileId := line.FileIdentifier
+			log.Attributes[types.AttributeDatabaseName] = tailer.database
+			log.Attributes[types.AttributeTableName] = tailer.table
+
+			for k, v := range tailer.resources {
+				log.Resource[k] = v
+			}
+
+			successfulParse := false
+			for _, parser := range tailer.logLineParsers {
+				err := parser.Parse(log)
+				if err == nil {
+					successfulParse = true
+					break // successful parse
+				} else if logger.IsDebug() {
+					logger.Debugf("readLines: parser error for filename %q: %v", tailer.tail.Filename, err)
+				}
+			}
+
+			if successfulParse {
+				// Successful parse, remove the raw message
+				delete(log.Body, types.BodyKeyMessage)
+			}
+
+			// Write after parsing to ensure these values are always set to values we need for acking.
+			log.Attributes[cursor_position] = position
+			log.Attributes[cursor_file_id] = currentFileId
+			log.Attributes[cursor_file_name] = tailer.tail.Filename
+
+			outputQueue <- log
+		}
+	}
+}
```

Comment:
I see this function is very permissive in its error handling, so my previous comments are maybe not warranted, though I suppose the danger is that a future maintainer doesn't quite follow the same implicit permissive error handling in this function. I'm not really sure how to address making explicit the principal that this function should not abate unless the intent is to abandon the file completely.